README.md
# Now we test

[![Build Status](https://travis-ci.org/lucasconstantino/now-we-test.svg?branch=master)](https://travis-ci.org/lucasconstantino/now-we-test)
[![coverage](https://img.shields.io/codecov/c/github/lucasconstantino/now-we-test.svg?style=flat-square)](https://codecov.io/github/lucasconstantino/now-we-test)
[![oclif](https://img.shields.io/badge/cli-oclif-brightgreen.svg)](https://oclif.io)
[![npm version](https://img.shields.io/npm/v/now-we-test.svg?style=flat-square)](https://www.npmjs.com/package/now-we-test)

Temporary lib to help test [Now.sh](https://zeit.co/now) lambdas written in Node.js. It was created with the purpose to eventually be deprecated when [`now dev` command](https://github.com/zeit/now-cli/issues/1681) is released.

## Purpose

Help unit test and locally serve lambdas written to deploy using Now.

## Installation

`yarn add now-we-test --dev`

## Usage

### For unit tests

In your unit tests, you can use the `simulator` helper to simulate a lambda execution environment without needing a running server:

```js
import { simulator } from 'now-we-test'
import func from '../your-lambda-implementation'

const app = simulator(func)

describe('func', () => {
  it('should do something', async () => {
    const result = await app.get('/')

    expect(result.text).toBe('[the value returned by the lambda]')
  })
})
```

Signature:

```
simulator(
  fn: Function,
  [builder="@now/node"]: String
)
```

Alternatively to the default `@now/node` building evironment, you can also simulate a [`now-micro`](https://github.com/lucasconstantino/now-micro) passing it as second parameter.

> The result of the invokation of `simulator` helper is a [SuperTest](https://github.com/visionmedia/supertest) instance, so please refer to that documentation for details on possible assertions and usage.

### For locally serving lambdas

_Now we test_ provides a simple CLI to serve your lambdas locally - useful for testing integration with other tools.

> Keep in mind there is only so much it can do in terms of simulation, so you should never completelly trust what works locally would necessarily work on Now.sh environment.

#### Commands

- [`now-we-test serve`](#now-we-test-serve)
- [`now-we-test help [COMMAND]`](#now-we-test-help-command)

##### `now-we-test serve`

Serves the lambdas locally

```
USAGE
  $ now-we-test serve [PATH]

ARGUMENTS
  PATH  [default: .] The path to the directory where now.json stands

OPTIONS
  -f, --format=format  [default: combined] logger format (morgan compliant)
  -p, --port=port      [default: 3000] port to use
```

_See code: [lib/commands/serve.js](https://github.com/lucasconstantino/now-we-test/blob/v0.0.0/src/commands/serve.js)_

> Only `@now/node`, `@now/static`, and `now-micro` builders are available so far.

##### `now-we-test help [COMMAND]`

display help for a now-we-test command

```
USAGE
  $ now-we-test help [COMMAND]

ARGUMENTS
  COMMAND  command to show help for

OPTIONS
  --all  see all commands in CLI
```

_See code: [@oclif/plugin-help](https://github.com/oclif/plugin-help/blob/v2.1.4/src/commands/help.ts)_


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.9.0"></a>
# [0.9.0](https://github.com/lucasconstantino/now-we-test/compare/v0.8.1...v0.9.0) (2019-01-30)


### Features

* updated simulator to handle sent body ([37640a3](https://github.com/lucasconstantino/now-we-test/commit/37640a3))



<a name="0.8.1"></a>
## [0.8.1](https://github.com/lucasconstantino/now-we-test/compare/v0.8.0...v0.8.1) (2019-01-30)



<a name="0.8.0"></a>
# [0.8.0](https://github.com/lucasconstantino/now-we-test/compare/v0.6.1...v0.8.0) (2019-01-08)


### Bug Fixes

* adapted simulator helper to new lambda agnostic builder simulation ([333726c](https://github.com/lucasconstantino/now-we-test/commit/333726c))
* remove url cleaning from static handler ([7ae5b80](https://github.com/lucasconstantino/now-we-test/commit/7ae5b80))


### Features

* added [@now](https://github.com/now)/static support ([97514af](https://github.com/lucasconstantino/now-we-test/commit/97514af))
* updated simulators to be lambda agnostic ([e324794](https://github.com/lucasconstantino/now-we-test/commit/e324794))



<a name="0.7.0"></a>
# [0.7.0](https://github.com/lucasconstantino/now-we-test/compare/v0.6.1...v0.7.0) (2019-01-08)


### Bug Fixes

* adapted simulator helper to new lambda agnostic builder simulation ([333726c](https://github.com/lucasconstantino/now-we-test/commit/333726c))


### Features

* added [@now](https://github.com/now)/static support ([97514af](https://github.com/lucasconstantino/now-we-test/commit/97514af))
* updated simulators to be lambda agnostic ([e324794](https://github.com/lucasconstantino/now-we-test/commit/e324794))



<a name="0.6.1"></a>
## [0.6.1](https://github.com/lucasconstantino/now-we-test/compare/v0.6.0...v0.6.1) (2019-01-08)



<a name="0.6.0"></a>
# [0.6.0](https://github.com/lucasconstantino/now-we-test/compare/v0.5.0...v0.6.0) (2019-01-05)


### Features

* added CORS support to serve command ([2aa6dcc](https://github.com/lucasconstantino/now-we-test/commit/2aa6dcc))



<a name="0.5.0"></a>
# [0.5.0](https://github.com/lucasconstantino/now-we-test/compare/v0.4.0...v0.5.0) (2019-01-05)


`serve` command now behaves more similarly to the Now 2.0 @now/node environment, but provides full Micro environment using the [now-micro builder](https://github.com/lucasconstantino/now-micro).

### Bug Fixes

* lint warnings ([9feef0d](https://github.com/lucasconstantino/now-we-test/commit/9feef0d))


### Features

* added simulator help to replace "lambda" but with multiple builders compatibility ([babe13a](https://github.com/lucasconstantino/now-we-test/commit/babe13a))
* augmented cache cleanup to basically include everything (total refresh) ([923976d](https://github.com/lucasconstantino/now-we-test/commit/923976d))
* simulated [@now](https://github.com/now)/node thrown error handling ([f24cbb6](https://github.com/lucasconstantino/now-we-test/commit/f24cbb6))
* updated serve command to use simulators ([2a1dbbc](https://github.com/lucasconstantino/now-we-test/commit/2a1dbbc))

### Deprecated

* lambda function is deprecated and will be removed soon



<a name="0.4.0"></a>
# 0.4.0 (2018-12-31)


### Bug Fixes

* fixed node < 7 ([3a36cd2](https://github.com/lucasconstantino/now-we-test/commit/3a36cd2))
* fixed route matching for node < 10 ([928cff8](https://github.com/lucasconstantino/now-we-test/commit/928cff8))
* fixed serving matching urls to be strict, not loose ([ca4dfb3](https://github.com/lucasconstantino/now-we-test/commit/ca4dfb3))
* use same regex capturing group syntax as now.sh expects ([e6a85ba](https://github.com/lucasconstantino/now-we-test/commit/e6a85ba))


### Features

* added lambda helper implementaiton and tests ([4520bc1](https://github.com/lucasconstantino/now-we-test/commit/4520bc1))
* implemented lambda serving command ([eb1cb9a](https://github.com/lucasconstantino/now-we-test/commit/eb1cb9a))
* updated serve command to respect methods ([606f03c](https://github.com/lucasconstantino/now-we-test/commit/606f03c))
* updated serve command to return http server for tests and outer control ([07736c3](https://github.com/lucasconstantino/now-we-test/commit/07736c3))


tests/lambda.test.js
const { lambda } = require('now-we-test')

describe('lambda', () => {
  const lambdas = {
    simple: jest.fn(() => 'Hello world!'),
    async: jest.fn(async () => 'Hello world!'),
    responding: jest.fn((req, res) => res.end('Hello world!')),
  }

  beforeEach(() => jest.clearAllMocks())

  describe('constructor', () => {
    it('should be a function', () => {
      expect(lambda).toBeInstanceOf(Function)
    })

    it('should throw when not invoked with a Function', () => {
      expect(() => {
        lambda()
      }).toThrow('Must be called with a function')
    })

    it('should return a supertest instance', () => {
      const app = lambda(() => {})

      expect(app).toHaveProperty('get')
      expect(app).toHaveProperty('post')
      expect(app).toHaveProperty('put')
      expect(app).toHaveProperty('delete')
    })
  })

  describe('simple', () => {
    it('should run a simple value returning lambda', async () => {
      const app = lambda(lambdas.simple)
      const result = await app.get('/')

      expect(lambdas.simple).toHaveBeenCalledTimes(1)
      expect(result).toHaveProperty('status', 200)
      expect(result).toHaveProperty('text', 'Hello world!')
    })
  })

  describe('async', () => {
    it('should run an async value returning lambda', async () => {
      const app = lambda(lambdas.async)
      const result = await app.get('/')

      expect(lambdas.async).toHaveBeenCalledTimes(1)
      expect(result).toHaveProperty('status', 200)
      expect(result).toHaveProperty('text', 'Hello world!')
    })
  })

  describe('responding', () => {
    it('should run an response dispatching lambda', async () => {
      const app = lambda(lambdas.async)
      const result = await app.get('/')

      expect(lambdas.async).toHaveBeenCalledTimes(1)
      expect(result).toHaveProperty('status', 200)
      expect(result).toHaveProperty('text', 'Hello world!')
    })
  })
})


tests/simulator.test.js
const console = require('console-suppress').default
const { simulator } = require('now-we-test')

const lambdas = {
  returning: jest.fn(() => 'Hello world!'),
  responding: jest.fn((req, res) => {
    res.end('Hello world!')
  }),

  asyncReturning: jest.fn(async () => 'Hello world!'),
  asyncResponding: jest.fn(async (req, res) => {
    res.end('Hello world!')
  }),

  throwing: jest.fn(() => {
    throw new Error('Ooops!')
  })
}

describe('simulator', () => {
  beforeEach(() => jest.clearAllMocks())
  beforeEach(() => console.cleanSuppressors())

  describe('constructor', () => {
    it('should be a function', () => {
      expect(() => simulator()).toBeInstanceOf(Function)
    })

    it('should throw when not invoked with a Function', () => {
      expect(() => simulator()).toThrow('Must be called with a function')
    })

    it('should return a supertest instance', () => {
      const app = simulator(() => {})

      expect(app).toHaveProperty('get')
      expect(app).toHaveProperty('post')
      expect(app).toHaveProperty('put')
      expect(app).toHaveProperty('delete')
    })

    it('should throw when invoked with unknown builder', () => {
      expect(() => simulator(() => {}, 'unknown-builders')).toThrow(
        'Unsupported builder unknown-builders'
      )
    })
  })

  describe('@now/node', () => {
    describe('returning', () => {
      it('should timeout on a value returning lambda', async () => {
        await expect(
          simulator(lambdas.returning, '@now/node')
            .get('/')
            .timeout(100)
        ).rejects.toThrow(/Timeout of 100ms exceeded/)

        expect(lambdas.returning).toHaveBeenCalledTimes(1)
      })
    })

    describe('responding', () => {
      it('should run an response dispatching lambda', async () => {
        const app = simulator(lambdas.responding, '@now/node')
        const result = await app.get('/').timeout(100)

        expect(lambdas.responding).toHaveBeenCalledTimes(1)
        expect(result).toHaveProperty('status', 200)
        expect(result).toHaveProperty('text', 'Hello world!')
      })
    })

    describe('async', () => {
      it('should timeout an async value returning lambda', async () => {
        await expect(
          simulator(lambdas.asyncReturning, '@now/node')
            .get('/')
            .timeout(100)
        ).rejects.toThrow(/Timeout of 100ms exceeded/)

        expect(lambdas.asyncReturning).toHaveBeenCalledTimes(1)
      })

      it('should run an async response dispatching lambda', async () => {
        const app = simulator(lambdas.asyncResponding, '@now/node')
        const result = await app.get('/').timeout(100)

        expect(lambdas.asyncResponding).toHaveBeenCalledTimes(1)
        expect(result).toHaveProperty('status', 200)
        expect(result).toHaveProperty('text', 'Hello world!')
      })
    })

    describe('throwing', () => {
      it('should return 502 on a throwing lambda', async () => {
        const app = simulator(lambdas.throwing, '@now/node')
        const result = await app.get('/').timeout(100)

        expect(lambdas.throwing).toHaveBeenCalledTimes(1)
        expect(result).toHaveProperty('status', 502)
        expect(result.text).toMatch(/NO_STATUS_CODE_FROM_LAMBDA/)
      })
    })
  })

  describe('now-micro', () => {
    describe('returning', () => {
      it('should timeout on a value returning lambda', async () => {
        const app = simulator(lambdas.returning, 'now-micro')
        const result = await app.get('/').timeout(100)

        expect(lambdas.returning).toHaveBeenCalledTimes(1)
        expect(result).toHaveProperty('status', 200)
        expect(result).toHaveProperty('text', 'Hello world!')
      })
    })

    describe('responding', () => {
      it('should run an response dispatching lambda', async () => {
        const app = simulator(lambdas.responding, 'now-micro')
        const result = await app.get('/').timeout(100)

        expect(lambdas.responding).toHaveBeenCalledTimes(1)
        expect(result).toHaveProperty('status', 200)
        expect(result).toHaveProperty('text', 'Hello world!')
      })
    })

    describe('async', () => {
      it('should timeout an async value returning lambda', async () => {
        const app = simulator(lambdas.asyncReturning, 'now-micro')
        const result = await app.get('/').timeout(100)

        expect(lambdas.asyncReturning).toHaveBeenCalledTimes(1)
        expect(result).toHaveProperty('status', 200)
        expect(result).toHaveProperty('text', 'Hello world!')
      })

      it('should run an async response dispatching lambda', async () => {
        const app = simulator(lambdas.asyncResponding, 'now-micro')
        const result = await app.get('/').timeout(100)

        expect(lambdas.asyncResponding).toHaveBeenCalledTimes(1)
        expect(result).toHaveProperty('status', 200)
        expect(result).toHaveProperty('text', 'Hello world!')
      })
    })

    describe('throwing', () => {
      it('should return 502 on a throwing lambda', async () => {
        console.error.suppress('Ooops!')

        const app = simulator(lambdas.throwing, 'now-micro')
        const result = await app.get('/').timeout(100)

        expect(lambdas.throwing).toHaveBeenCalledTimes(1)
        expect(result).toHaveProperty('status', 500)
        expect(result).toHaveProperty('text', 'Internal Server Error')
      })
    })
  })
})


tests/commands/serve.test.js
/* eslint-disable global-require */
const net = require('net')
const path = require('path')
const supertest = require('supertest')
const console = require('console-suppress').default
const ServeCommand = require('now-we-test/commands/serve')

const isPortAvailable = port =>
  new Promise((resolve, reject) => {
    const tester = net
      .createServer()
      .once('error', err =>
        err.code === 'EADDRINUSE' ? resolve(false) : reject(err)
      )
      .once('listening', () =>
        tester.once('close', () => resolve(true)).close()
      )
      .listen(port)
  })

describe('commands', () => {
  beforeEach(() => console.cleanSuppressors())

  describe('serve', () => {
    let app

    afterEach(done => (app && app.listening ? app.close(done) : done()))

    describe('@now/static', () => {
      const basePath = path.resolve(__dirname, '../fixtures/now-static-project')

      it('should return static file content', async () => {
        app = await ServeCommand.run([basePath])

        await supertest(app)
          .get('/README.md')
          .expect(200, '> This is a sample static file for testing purposes.\n')
      })
    })

    describe('@now/node', () => {
      const basePath = path.resolve(__dirname, '../fixtures/now-node-project')

      const lambdas = {
        unrouted: require('../fixtures/now-node-project/lambda.js'),
        returning: require('../fixtures/now-node-project/lambdas/returning'),
        asyncReturning: require('../fixtures/now-node-project/lambdas/asyncReturning'),
        responding: require('../fixtures/now-node-project/lambdas/responding'),
        throwing: require('../fixtures/now-node-project/lambdas/throwing')
      }

      beforeEach(() => {
        Object.keys(lambdas).forEach(name => {
          lambdas[name].mockClear()
        })
      })

      describe('general serve tests', () => {
        it('should have a run method', () => {
          expect(ServeCommand.run).toBeInstanceOf(Function)
        })

        it('should throw when no now.json is found', async () => {
          await expect(ServeCommand.run([])).rejects.toThrow(
            /No now\.json found/
          )
        })

        it('should start serving when working now.json found', async () => {
          app = await ServeCommand.run([basePath])

          expect(app.listening).toBe(true)
        })

        it('should listen on specific port when configured', async () => {
          await expect(isPortAvailable(3001)).resolves.toBe(true)
          app = await ServeCommand.run([basePath, '--port=3001'])
          await expect(isPortAvailable(3001)).resolves.toBe(false)
        })
      })

      it('should return 404 when no lambda found on the requested URL', async () => {
        console.error.suppress(/No input matching requested path/)

        app = await ServeCommand.run([basePath])

        await supertest(app)
          .get('/non/existing/path')
          .expect(404, 'No input matching requested path')
      })

      describe('returning', () => {
        it('should timeout on a value returning lambda', async () => {
          app = await ServeCommand.run([basePath])

          await expect(
            supertest(app)
              .get('/custom/path/returning.js')
              .timeout(100)
          ).rejects.toThrow(/Timeout of 100ms exceeded/)

          expect(lambdas.returning).toHaveBeenCalledTimes(1)
        })
      })

      describe('async', () => {
        it('should timeout on an async value returning lambda', async () => {
          app = await ServeCommand.run([basePath])

          await expect(
            supertest(app)
              .get('/custom/path/asyncReturning.js')
              .timeout(100)
          ).rejects.toThrow(/Timeout of 100ms exceeded/)

          expect(lambdas.asyncReturning).toHaveBeenCalledTimes(1)
        })
      })

      describe('responding', () => {
        it('should run an response dispatching lambda', async () => {
          app = await ServeCommand.run([basePath])

          await supertest(app)
            .get('/custom/path/responding.js')
            .expect(200, 'Hello world!')

          expect(lambdas.responding).toHaveBeenCalledTimes(1)
        })
      })

      describe('throwing', () => {
        it('should run a throwing lambda', async () => {
          app = await ServeCommand.run([basePath])

          await supertest(app)
            .get('/custom/path/throwing.js')
            .expect(502, /NO_STATUS_CODE_FROM_LAMBDA/)

          expect(lambdas.throwing).toHaveBeenCalledTimes(1)
        })
      })

      it('should run on unrouted (direct path) lambdas', async () => {
        app = await ServeCommand.run([basePath])

        await supertest(app)
          .get('/lambda.js')
          .expect(200, 'Hello world!')

        expect(lambdas.unrouted).toHaveBeenCalledTimes(1)
      })

      it('should respect provided methods', async () => {
        app = await ServeCommand.run([basePath])

        await supertest(app)
          .get('/method-path')
          .expect(200, 'Hello world!')

        expect(lambdas.unrouted).toHaveBeenCalledTimes(1)

        console.error.suppress(/No input matching requested path/)

        await supertest(app)
          .post('/method-path')
          .expect(404, 'No input matching requested path')

        expect(lambdas.unrouted).toHaveBeenCalledTimes(1)
      })

      it('should work with POST requests', async () => {
        app = await ServeCommand.run([basePath])

        await supertest(app)
          .post('/post-path')
          .send({ name: 'John' })
          .timeout(1000)
          .expect(200, 'Hello John!')

        expect(lambdas.unrouted).toHaveBeenCalledTimes(1)
      })

      it('should accept any method on routes without method restriction', async () => {
        app = await ServeCommand.run([basePath])

        await supertest(app).get('/no-method-path').expect(200, 'Hello world!')
        await supertest(app).post('/no-method-path').expect(200, 'Hello world!')
        await supertest(app).put('/no-method-path').expect(200, 'Hello world!')
        await supertest(app).delete('/no-method-path').expect(200, 'Hello world!')
        await supertest(app).options('/no-method-path').expect(200, 'Hello world!')

        expect(lambdas.unrouted).toHaveBeenCalledTimes(5)
      })
    })

    describe('now-micro', () => {
      const basePath = path.resolve(__dirname, '../fixtures/now-micro-project')

      const lambdas = {
        unrouted: require('../fixtures/now-micro-project/lambda.js'),
        returning: require('../fixtures/now-micro-project/lambdas/returning'),
        asyncReturning: require('../fixtures/now-micro-project/lambdas/asyncReturning'),
        responding: require('../fixtures/now-micro-project/lambdas/responding'),
        throwing: require('../fixtures/now-micro-project/lambdas/throwing')
      }

      beforeEach(() => {
        Object.keys(lambdas).forEach(name => {
          lambdas[name].mockClear()
        })
      })

      describe('returning', () => {
        it('should run a value returning lambda', async () => {
          app = await ServeCommand.run([basePath])

          await supertest(app)
            .get('/custom/path/returning.js')
            .expect(200, 'Hello world!')

          expect(lambdas.returning).toHaveBeenCalledTimes(1)
        })
      })

      describe('async', () => {
        it('should run an async value returning lambda', async () => {
          app = await ServeCommand.run([basePath])

          await supertest(app)
            .get('/custom/path/asyncReturning.js')
            .expect(200, 'Hello world!')

          expect(lambdas.asyncReturning).toHaveBeenCalledTimes(1)
        })
      })

      describe('responding', () => {
        it('should run an response dispatching lambda', async () => {
          app = await ServeCommand.run([basePath])

          await supertest(app)
            .get('/custom/path/responding.js')
            .expect(200, 'Hello world!')

          expect(lambdas.responding).toHaveBeenCalledTimes(1)
        })
      })

      describe('throwing', () => {
        it('should run a throwing lambda', async () => {
          console.error.suppress('Ooops!')

          app = await ServeCommand.run([basePath])

          await supertest(app)
            .get('/custom/path/throwing.js')
            .expect(500, 'Internal Server Error')

          expect(lambdas.throwing).toHaveBeenCalledTimes(1)
        })
      })

      it('should run on unrouted (direct path) lambdas', async () => {
        app = await ServeCommand.run([basePath])

        await supertest(app)
          .get('/lambda.js')
          .expect(200, 'Hello world!')

        expect(lambdas.unrouted).toHaveBeenCalledTimes(1)
      })

      it('should respect provided methods', async () => {
        app = await ServeCommand.run([basePath])

        await supertest(app)
          .get('/method-path')
          .expect(200, 'Hello world!')

        expect(lambdas.unrouted).toHaveBeenCalledTimes(1)

        console.error.suppress(/No input matching requested path/)

        await supertest(app)
          .post('/method-path')
          .expect(404, 'No input matching requested path')

        expect(lambdas.unrouted).toHaveBeenCalledTimes(1)
      })
    })
  })
})


lib/simulator.js
const path = require('path')
const http = require('http')
const micro = require('micro')
const supertest = require('supertest')
const invariant = require('invariant')
const staticHandler = require('serve-handler')
const toPromise = require('stream-to-promise')

const lambdaSimulator = createApp => ({ entrypoint }) => async (req, res) => {
  const lambda =
    typeof entrypoint === 'string' ? require(entrypoint) : entrypoint

  invariant(typeof lambda === 'function', 'Lambdas must be functions')

  const method = req.method.toLowerCase()
  const app = createApp(lambda)
  const agent = supertest(app)
  const body = (await toPromise(req)).toString()

  // @TODO: should simulate more complex requests
  const response = await agent[method](req.url)
    .set(req.headers)
    .send(body)

  const { headers, text, statusCode } = response

  // @TODO: should simulate more complex responses

  // proxy headers
  Object.keys(headers).forEach(name => {
    res.setHeader(name, headers[name])
  })

  // proxy statusCode
  res.statusCode = statusCode

  // proxy text & end response
  res.end(text)
}

module.exports.simulators = {
  '@now/static': ({ entrypoint, rootPath, pathname }) => (req, res) => {
    // force static url path
    req.url = path.relative(process.cwd(), entrypoint)

    staticHandler(req, res, {
      cleanUrls: false
    })
  },

  /**
   * @node/node builder simulator.
   */
  '@now/node': lambdaSimulator(lambda =>
    http.createServer((req, res) => {
      try {
        lambda(req, res)
      }
      catch (err) {
        res.statusCode = 502
        res.end(`
An error occurred with your deployment

NO_STATUS_CODE_FROM_LAMBDA
`)
      }
    })
  ),

  /**
   * now-micro builder simulator.
   */
  'now-micro': lambdaSimulator(micro)
}

/**
 * Lambda simulator
 */
module.exports.simulator = (entrypoint, builder = '@now/node') => {
  invariant(
    module.exports.simulators[builder],
    `Unsupported builder ${builder}`
  )

  invariant(typeof entrypoint === 'function', 'Must be called with a function')

  return supertest(module.exports.simulators[builder]({ entrypoint }))
}

// retro-compatibility.
module.exports.lambda = module.exports.simulator


lib/commands/serve.js
/* eslint-disable import/no-dynamic-require,global-require */
const { existsSync } = require('fs')
const path = require('path')
const { Command, flags } = require('@oclif/command')
const minimatch = require('minimatch')
const invariant = require('invariant')
const micro = require('micro')
const Regex = require('named-regexp-groups')
const morgan = require('morgan')
const clearModule = require('clear-module')
const { simulators } = require('../simulator')

const { createError } = micro

const builders = Object.keys(simulators)

// FIXME: Supporting canary version of node builders until Typescript will make it to the stable release (https://github.com/zeit/now-builders/issues/46)
builders.push('@now/node@canary')

class ServeCommand extends Command {
  async run () {
    const {
      flags: { port, format, timeout },
      args: { path: basePath }
    } = this.parse(ServeCommand)
    // const name = flags.name || 'world'

    const rootPath = path.resolve(process.cwd(), basePath)
    const configPath = path.resolve(rootPath, 'now.json')

    invariant(existsSync(configPath), `No now.json found at ${configPath}`)

    this.debug(`using ${configPath}`)

    const { builds = [], routes = [] } = require(configPath)

    // find all supported builders
    const supportedBuilds = builds.filter(
      ({ use }) => builders.indexOf(use) > -1
    )

    invariant(supportedBuilds.length > 0, 'No supported builds found')

    // preconfigure formatter
    const httpLogger = morgan(format)

    const app = micro(async (req, res, ...args) => {
      const paths = [req.url]

      if (routes.length) {
        const route = routes.find(
          ({ src, methods }) =>
            (!methods || methods.indexOf(req.method) !== -1) &&
            req.url.match(new RegExp(`^${src.replace(/\(\?<[a-z]+>/, '(')}$`))
        )

        // compatible route found
        if (route) {
          const captGroups = []

          // see https://stackoverflow.com/a/11443943
          route.src.replace(/\(\?<([a-z]+)>/gi, (match, id) => {
            captGroups.push(id)
          })

          // register route destination
          // @TODO: should normalize dest to PCRE complient backreferences
          // see: http://perldoc.perl.org/perlretut.html#Named-backreferences
          paths.push(
            req.url.replace(
              new Regex(route.src),
              captGroups.reduce(
                (dest, id) => dest.replace(new RegExp('\\$' + id), `$+{${id}}`),
                route.dest
              )
            )
          )
        }
      }

      const pathname = paths
        // remove trailing slash
        .map(url => url.replace(/^\//, ''))
        // remove query params
        .map(url => url.replace(/\?.*/, ''))
        // remove potential duplicated extension (i.e. .js.js or .ts.ts)
        .map(url => url.replace(/(\.[tj]s)\1$/, '$1'))

        // find the first build registered url
        .find(url => supportedBuilds.find(({ src }) => minimatch(url, src)))

      if (!pathname) {
        throw createError(404, 'No input matching requested path')
      }

      // find the compatible builder
      const builder = supportedBuilds.find(({ src }) =>
        minimatch(pathname, src)
      )

      // FIXME: remove once @now/node stable will support Typescript
      // removing unnecessary canary tag to match simulator key
      builder.use = builder.use.replace('@canary', '')

      invariant(builder, `No builder found for path ${pathname}`)

      const targetPath = path.resolve(rootPath, pathname)

      // handle not-found paths.
      if (!existsSync(targetPath)) {
        throw createError(404, `Expected input not found for path ${pathname}`)
      }

      // ensure every module state and context is cleaned-up between requests.
      clearModule.match(/.*/)

      httpLogger(req, res, err => {
        if (err) throw err

        const context = {
          entrypoint: targetPath,
          pathname: `/${pathname}`,
          rootPath,
          builder,
          server: app,
          timeout
        }

        simulators[builder.use](context)(req, res)
      })
    })

    app.listen(port, () =>
      this.log(`Serving lambdas at http://localhost:${port}`)
    )

    // save it to make it possible to bring it down.
    return app
  }
}

// ServeCommand.description = `Serves the lambdas locally
// ...
// Extra documentation goes here
// `

ServeCommand.description = 'Serves the lambdas locally'

ServeCommand.args = [
  {
    name: 'path',
    required: false,
    description: 'The path to the directory where now.json stands',
    default: '.'
  }
]

ServeCommand.flags = {
  port: flags.string({
    char: 'p',
    description: 'port to use',
    default: 3000,
    env: 'PORT'
  }),
  format: flags.string({
    char: 'f',
    description: 'logger format (morgan compliant)',
    default: 'combined'
  }),
  timeout: flags.string({
    char: 't',
    description: 'general request timeout',
    default: null
  })
}

module.exports = ServeCommand
